src/storage/src/hummock/state_store.rs

```diff
@@ -75,28 +76,39 @@ impl HummockStorage {
         epoch: u64,
         read_options: ReadOptions,
     ) -> StorageResult<StreamTypeOfIter<HummockStorageIterator>> {
-        let read_version_tuple =
-            self.build_read_version_tuple(epoch, read_options.table_id, &key_range)?;
+        let read_version_tuple = self
+            .build_read_version_tuple(epoch, read_options.table_id, &key_range)
+            .await?;
 
         self.hummock_version_reader
             .iter(key_range, epoch, read_options, read_version_tuple)
             .await
     }
 
-    fn build_read_version_tuple(
+    async fn build_read_version_tuple(
         &self,
         epoch: u64,
         table_id: TableId,
         key_range: &TableKeyRange,
     ) -> StorageResult<(Vec<ImmutableMemtable>, Vec<SstableInfo>, CommittedVersion)> {
         let pinned_version = self.pinned_version.load();
-        validate_epoch(pinned_version.safe_epoch(), epoch)?;
-
+        let mut committed_version = (**pinned_version).clone();
+        if epoch < pinned_version.safe_epoch() {
+            // try to query hummock version from backup
+            if let Some(backup_version) = self.backup_reader.try_get_hummock_version(epoch).await {
```

Comment:
Agree with the benefit of "clearly identify misusage or potential bugs ". However, that also means the frontend need to be aware of "reading from backup", which is what the current implementation aims to avoid. But if we will support "reading from backup for dropped table", then frontend has to be aware of it anyway.

Reply:
Refactored.